Stabilize `feature(trait_upcasting)`

WaffleLapkin opened the pull request:
This feature was "done" for a while now, I think it's finally time to stabilize it! Stabilization report: https://github.com/rust-lang/rust/pull/134367#issuecomment-2545839841.
cc reference PR: https://github.com/rust-lang/reference/pull/1622.

Closes #65991 (tracking issue), closes #89460 (the lint is no longer future incompat).

r? compiler-errors

WaffleLapkin committed:
stabilize `feature(trait_upcasting)`

WaffleLapkin committed:
change `deref_into_dyn_supertrait` lint wording

(so that it doesn't talk about trait upcasting stabilization in the
future tense)

WaffleLapkin committed:
remove `feature(trait_upcasting)` from tests and bless them

WaffleLapkin committed:
remove unstable book entry for `feature(trait_upcasting)`

compiler-errors commented:
Could you please provide some links that specify what has been done since the last stabilization? Ideally both the issues and the resulting fix PRs.

Ty for doing this tho :) 

Gna nominate for T-lang but this should be an easy decision.

compiler-errors commented:
And maybe link to the last stabilization too

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12345301033/job/34449189751) [(plain)](https://github.com/rust-lang/rust/commit/a5e3c914fa30993f109ca818b533c45e89f4c7ff/checks/34449189751/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

WaffleLapkin committed:
remove use of `feature(trait_upcasting)` from miri

WaffleLapkin committed:
remove use of `feature(trait_upcasting)` from core tests

WaffleLapkin commented:
## Stabilization report

The core of the feature hasn't changed since the [last attempt to stabilize it](https://github.com/rust-lang/rust/issues/65991#issuecomment-1670127881).

But as a reminder this feature allows "upcasting" `dyn Sub` to `dyn Super`, such as in this example:

```rust
trait Sub: Super {}
trait Super {}

fn upcast(x: &dyn Sub) -> &dyn Super {
    x // implicit coercion
}
```

This is a long wanted feature that people used workarounds for for a long while now.

The only possible downside is that this forces us into including more data in the vtables. However, our measurements show that the overhead is mostly negligible. Also note that we ate already including this overhead on stable for countless versions and no one ever complained.

This feature also locks us into requiring that metadata of a pointer must be "valid" (i.e. it's not _safe_ to create a `*const dyn Sub` with a dangling vtable). This seems to be generally agreed upon, so I see no issue here.

I believe that the feature is well tested and is ready for stabilization.

### Previous stabilization attempt problems

After the last attempt to stabilize this feature @steffahn found two unsound interactions between trait upcasting and pointer casting (one of which also required `feature(arbitrary_self_types)`): #120222 and #120217. This caused a revert of the stabilization PR.

Both issues were since fixed in #120248 by adding additional checks for casting pointers, to uphold the soundness requirements of pointers to trait objects.

No new issues were found since.

WaffleLapkin commented:
(@RalfJung I'm pretty sure https://github.com/rust-lang/rust/labels/I-lang-easy-decision implies nomination and you don't need https://github.com/rust-lang/rust/labels/I-lang-nominated)

RalfJung commented:
@traviscross told me the opposite -- it still needs to be nominated with the regular label.

RalfJung commented:
> The only possible downside 

That's a very strong statement.^^

I'd say there is at least one other downside, or point worth mentioning: this allows upcasting of *raw* trait pointers in safe code. That puts constraints on their safety invariant -- specifically, even `*const dyn Trait` must always come with a vtable that is valid for `Trait`.

WaffleLapkin commented:
@RalfJung I did call this out, but maybe not explicitly enough ("the only" slipped from an earlier draft). I've updated the wording to more clearly highlight this.

nikomatsakis commented:
> I'd say there is at least one other downside, or point worth mentioning: this allows upcasting of _raw_ trait pointers in safe code. That puts constraints on their library invariant (safety invariant) -- specifically, even `*const dyn Trait` must always come with a vtable that is valid for `Trait`.
> 
> I'm fine with that, and AFAIK @rust-lang/types agrees, but it is a choice we are making here that should be called out explicitly. @rust-lang/opsem is still discussing what to do with the language invariant for raw trait pointers; having that be different from the library invariant is likely surprising but OTOH we still might want a weaker invariant here.

worth noting that we did formally decide this here: https://github.com/rust-lang/rust/issues/101336

nikomatsakis commented:
@rfcbot fcp merge

I'm so excited for this. It's going to make Salsa 3.0 work a *lot* better (along with a bunch of other code...). 

[Stabilization report](https://github.com/rust-lang/rust/pull/134367#issuecomment-2545839841)

compiler-errors commented:
@scottmcm:

```
fn foo(_1: &dyn Sub) -> &dyn Sup {
    debug x => _1;
    let mut _0: &dyn Sup;

    bb0: {
        _0 = copy _1 as &dyn Sup (PointerCoercion(Unsize, Implicit));
        return;
    }
}
```

They are a form of unsizing in both HIR and MIR.

WaffleLapkin commented on compiler/rustc_lint/src/deref_into_dyn_supertrait.rs:
I'm not actually sure what to do with this lint. "Future incompatible" doesn't make sense anymore, since this PR is actually adding the breaking change... should I make it into a normal lint? Should I make it deny-by-default?

jackh726 replied:
It might be time to do a crater run for switching it to deny-by-default.

The conservative case is to keep it warn and switch it later - I think that's okay.

jackh726 replied:
Particularly, I could very much imagine people having these impls around today precisely *because* trait upcasting isn't stable. And so switching it later might be much easier.

RalfJung replied:
> "Future incompatible" doesn't make sense anymore, since this PR is actually adding the breaking change

So we can't keep this a lint and proceed with compilation when it triggers?

RalfJung replied:
Oh wait this is a `FutureReleaseSemanticsChange`, I've never seen one of those... I see now. Yeah probably the lint should just removed, but Cc @rust-lang/lang to be sure. This is about https://github.com/rust-lang/rust/issues/89460.

WaffleLapkin replied:
> So we can't keep this a lint and proceed with compilation when it triggers?

We can! 

The code that this lint lints against is basically useless now, because the deref impl is "shadowed" by unsize coercion. 

We would just make it *not* future incompatible, since there is no future anymore, only present.

jackh726 replied:
Well, we could do three things:
1) Completely disallow such impls, since they are functionally useless. This would be a breaking change both today and after stabilization.
2) Always warn (or deny) if we see a shadowed impl, since they are useless but not actively harmful.
3) Just ignore impls, without warning. (Or keep an allow by default lint.)

So, if we want to eventually move to (1), then keeping this future compat lint is useful. If we want to eventually just stick with (2), then this lint shouldn't be a future compat lint, but just a regular one. For (3), the lint can be removed completely with stabilization.

RalfJung replied:
Do we have any precedent for a lint that triggers on "useless" impls?

compiler-errors replied:
Not afaict, but it doesn't seem special in any way to lint on an impl we know is (almost) always gonna be shadowed during coercion.

RalfJung replied:
Yeah it seems reasonable, I was just wondering.

If this PR is supposed to also add that lint, I think that requires another round of t-lang discussion + approval.

WaffleLapkin replied:
I think disallowing the impl is silly, it *can* be useful, for example in combination with genetic code. I'm fine with making this a normal lint or removing the line altogether.

WaffleLapkin replied:
Should I nominate for T-lang discussion again, since there doesn't seem to be a consensus of what to do?

jackh726 replied:
Yes, I think that's good

lukas-code replied:
FWIW we currently do not warn for `Deref` impls that are shadowed by other kinds of coercions, for example:

<details>
<summary>array->slice unsizing example</summary>

```rust
use std::ops::Deref;

struct Wrap<T: ?Sized>(T);

impl Deref for Wrap<[u8; 0]> {
    type Target = Wrap<[u8]>;

    fn deref(&self) -> &Self::Target {
        println!("deref!");
        self
    }
}

fn main() {
    let _: &Wrap<[u8]> = &Wrap([]); // doesn't print "deref!"
    let _: &Wrap<[u8]> = Deref::deref(&Wrap([])); // prints "deref!"
}
```

</details>

<details>
<summary>subtyping example</summary>

```rust
use std::ops::Deref;

struct Wrap<T: ?Sized>(T);

impl Deref for Wrap<fn(&())> {
    type Target = Wrap<fn(&'static ())>;

    fn deref(&self) -> &Self::Target {
        println!("deref!");
        self
    }
}

fn main() {
    let _: &Wrap<fn(&'static ())> = &Wrap((|_| ()) as fn(&())); // doesn't print "deref!"
    let _: &Wrap<fn(&'static ())> = Deref::deref(&Wrap((|_| ()) as fn(&()))); // prints "deref!"
}
```

</details>

So IMO if we keep this lint, it should apply to all `Deref` impls that are shadowed by any kind of coercion and not be arbitrarily restricted to the specific `Deref` impls that will no longer be called after this PR merges.

WaffleLapkin replied:
If we go that round I'd prefer to remove the lint & leave implementing the full `shadowed_deref` lint as a followup.

tmandry commented:
@rfcbot reviewed

(Already checked my box, but rfcbot doesn't seem to register this.)

WaffleLapkin commented:
I'm re-nominating for lang discussion, specifically to ask what to do with the `deref_into_dyn_supertrait` lint. As it is right now, it is a FCW lint that warns against impls of `Deref` for `dyn Trait` which will be shadowed by the trait upcasting (after this PR). I.e.

```rust
#![deny(deref_into_dyn_supertrait)]
#![allow(dead_code)]

use core::ops::Deref;

trait A {}
trait B: A {}
impl<'a> Deref for dyn 'a + B {
    type Target = dyn A;
    fn deref(&self) -> &Self::Target {
        todo!()
    }
}

fn take_a(_: &dyn A) {}

fn take_b(b: &dyn B) {
    take_a(b);
}
```
```
error: this `Deref` implementation is covered by an implicit supertrait coercion
 --> src/lib.rs:8:1
  |
8 | impl<'a> Deref for dyn 'a + B {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `dyn B` implements `Deref<Target = dyn A>` which conflicts with supertrait `A`
9 |     type Target = dyn A;
  |     -------------------- target type is a supertrait of `dyn B`
  |
  = warning: this will change its meaning in a future release!
  = note: for more information, see issue #89460 <https://github.com/rust-lang/rust/issues/89460>
```

The impl is not harmful, but is mostly useless, since it will no be called in coercions (after this PR). It _might_ in some cases be useful, if generic code expects `T: Deref`, although I'm doubting this actually happens. Another use case might be supporting old(er) compilers.

Either way the question is: **what to do with the lint?** The options are:

1. Completely disallow such impls, since they are functionally useless. This would be a breaking change both today and after stabilization.
2. Always warn (or deny) if we see a shadowed impl, since they are useless but not actively harmful.
    - i.e. turn this lint from an FCW to a normal warning
3. Just ignore impls, without warning. (Or keep an allow by default lint.)

I'm strongly in favor of options 2 and 3, since I do not see a reason to _forbid_ the impl -- it is not harmful (mostly just silly/useless). I'd probably keep the lint as a normal warning (warn by default), to highlight that it's shadowed by built in unsizing most of the time.

Veykril commented:
Dropping by to just say that rust-analyzer currently throws a type mismatch on trait upcasting coercions as the feature isn't implemented there yet https://github.com/rust-lang/rust-analyzer/issues/18083

WaffleLapkin commented:
Current status (so I don't forget): this is waiting on chalk/r-a properly handling upcasting and on T-lang to make a decision about the lint.

bors commented:
:umbrella: The latest upstream changes (presumably #133858) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

theemathas commented:
Trait upcasting has a soundness bug. Oops. #135315

steffahn commented:
I kind-of hate to ***once again*** be the one finding new issues at the last minute that could be further delaying this feature, sowwy 🫠

*but soundness is soundness!* 😇

nikomatsakis commented:
Better now than later. =)




On Thu, Jan 9, 2025, at 10:51 PM, Frank Steffahn wrote:
> 
> 
> I kind-of hate to **once again** be the one finding new issues at the last minute that could be further delaying this feature, sowwy 🫠
> 
> *but soundness is soundness!* 😇
> 
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/rust-lang/rust/pull/134367#issuecomment-2581688230>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AABF4ZTDY6O4L5GMLWW6EIT2J47TZAVCNFSM6AAAAABTVG5QT6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOBRGY4DQMRTGA>.
> You are receiving this because you were mentioned.Message ID: ***@***.***>
>

WaffleLapkin commented:
![image](https://github.com/user-attachments/assets/62cb6774-78fb-4f32-a54c-b0a25d715b2d)

But on a more serious note, thanks for finding the bugs (so we can fix them!). And also, it's nice to know about them before merging stabilization, so we don't have to revert anything. <3 @steffahn

WaffleLapkin commented:
Status update: I've implemented the feature in chalk/r-a.

This is now waiting on t-lang decision about the lint and some soundness fixes to the issues @steffahn found.

WaffleLapkin committed:
adjust a comment

We can't remove `TraitUpcasting` variant, since we need to use the index
in winnowing.

WaffleLapkin commented:
(sorry for the ping, I accidentally deleted the whole rustc dev guide; I reverted this, so yeah...)

compiler-errors commented:
Still waiting on those vtable fixes I think

@rustbot blocked

bors commented:
:umbrella: The latest upstream changes (presumably #135937) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

kellerkindt commented:
![Img_2025_01_31_19_06_20.jpeg](https://github.com/user-attachments/assets/bbac94ed-94fb-4e5a-87c3-3c6c0cc05343)

Someone probably needs to update the Labels. Thanks!

nikomatsakis commented:
We discussed the lint case in triage meeting. My current take is that we should remove the `deref_into_dyn_supertrait` lint. I agree with @WaffleLapkin that we can't make this a hard error for backwards compatibility reasons -- while I wish we *did* disallow impls for dyn types, that's complex, and not something we should do as part of this stabilization decision.

I don't believe anyone contends that the existence of this impl is "almost certainly a bug". I could lead to surprising behavior of the Deref impl does not do the same thing as the coercion, but that is generally a risk one carries when writing non-trivial `Deref` impls and doesn't seem particularly higher in this case.

That leaves the rationale as either avoiding user confusion or helping to avoid needless divergence (e.g., enforcing capitalization). I don't think either is especially strong in this case, this seems like a rather obscure situation, and (as @traviscross pointed out) it seems possible to confuse people in the other direction, as perhaps they want to have the impl and get misled into thinking there is something fundamentally wrong with it.

Given all this, I think our default Rust position for lints is "when in doubt, leave it out", therefore I am of the opinion we should just remove the lint.

compiler-errors commented:
Ok, I'd prefer if we keep the deref lint as `Allow` at least for impls. It can stay get ripped out of the trait solver.

@rustbot author

Otherwise LGTM, r=me after adjusting the lint allow level and any UI test fallout. Or if you'd like to remove it @WaffleLapkin, then I'm also open to that.

WaffleLapkin committed:
allow+update `deref_into_dyn_supertrait`

this commit makes `deref_into_dyn_supertrait` lint allow-by-default,
removes future incompatibility (we finally live in a broken world), and
changes the wording in the documentation.

previously documentation erroneously said that it lints against *usage*
of the deref impl, while it actually (since 104742) lints on the impl
itself (oooops, my oversight, should have updated it 2+ years ago...)

WaffleLapkin commented:
I personally think that the lint should be removed + reworked (as something like "shadowed deref impl"). But, for now I just made it allow-by-default + fixed up it's message and documentation.

WaffleLapkin commented:
With that, I think we can finally

![bors r plus](https://www.pietroalbini.org/bors-r-plus.png)
<!--
@bors r=compiler-errors
-->

bors commented:
:pushpin: Commit 1aac2f135aaa3c2ab75bdd945b7c0501097d22d0 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 1aac2f135aaa3c2ab75bdd945b7c0501097d22d0 -->
<!-- homu: {"type":"Approved","sha":"1aac2f135aaa3c2ab75bdd945b7c0501097d22d0","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13188171072/job/36815194426) [(plain)](https://github.com/rust-lang/rust/commit/1aac2f135aaa3c2ab75bdd945b7c0501097d22d0/checks/36815194426/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 27.5s done
#21 DONE 33.4s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
To only update this specific test, also pass `--test-args traits/trait-upcasting/mono-impossible.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/traits/trait-upcasting/mono-impossible.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/trait-upcasting/mono-impossible" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
--- stderr -------------------------------
warning: the feature `trait_upcasting` has been stable since 1.86.0-nightly and no longer requires an attribute to enable
##[warning]  --> /checkout/tests/ui/traits/trait-upcasting/mono-impossible.rs:3:12
   |
```

</details>

WaffleLapkin commented:
With that, I think we can finally

![bors r plus](https://www.pietroalbini.org/bors-r-plus.png)
<!--
@bors r=compiler-errors
-->

bors commented:
:pushpin: Commit 491599569c081985d6cc3eb4ab55d692e380e938 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 491599569c081985d6cc3eb4ab55d692e380e938 -->
<!-- homu: {"type":"Approved","sha":"491599569c081985d6cc3eb4ab55d692e380e938","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Kobzol commented:
Because the initial stabilization was later reverted: https://github.com/rust-lang/rust/pull/120233

gabrielecastellano commented:
> Because the initial stabilization was later reverted: #120233

Completely missed that. Now it is very clear :)
